File: app/src/system/saySystem.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____EventEmiter = require("src.core.EventEmiter")
local EventEmiter = ____EventEmiter.EventEmiter
____exports.SaySystem = __TS__Class()
local SaySystem = ____exports.SaySystem
SaySystem.name = "SaySystem"
__TS__ClassExtends(SaySystem, EventEmiter)
function SaySystem.prototype.init(self, app)
    self._app = app
end
function SaySystem.prototype.say(self, profile, text, options)
    if not self._app then
        return nil
    end
    local rs, err = table.unpack(self._app:call(
        "say",
        "say",
        profile,
        text,
        options
    ))
    if err then
        return nil
    end
    if options == nil then
        return nil
    end
    for ____, opt in ipairs(options) do
        if opt.name == rs then
            return opt
        end
    end
end
____exports.saySystem = __TS__New(____exports.SaySystem)
return ____exports

File: src/system/saySystem.ts
import { App } from "../core/App";
import { EventEmiter } from "../core/EventEmiter";

export type SayProfileStyle = "left" | "right"
export interface SayOption {
    title: string
    name: string
}

export interface ISayProfile {
    name: string
    title: string
    res: string
    style: SayProfileStyle
}


export class SaySystem extends EventEmiter {

    private _app?: App

    init(app: App): void {
        this._app = app
    }

    say(profile: ISayProfile, text: string, options?: SayOption[]): SayOption | undefined {
        if (!this._app) return undefined
        let [rs, err] = this._app.call("say", "say", profile, text, options)
        if (err) return undefined
        if (options === undefined) return undefined
        for (let opt of options) {
            if (opt.name == rs) return opt
        }
    }

}

export const saySystem = new SaySystem()